### ISSUE TITLE
Support fullscreen on iPhone for video elements

### ISSUE THREAD

**Issue by mishoboss** [dermotduffy/frigate-hass-card#1444]
I know iOS webview is not supporting the fullscreen API, but I'm thinking a custom JS implementation as a workaround - take the screen size, render a DIV with that size as an overlay and render the player inside that DIV.

What do you think about such an approach? Fullscreen on iOS is really missing.

**Comment by dermotduffy** [comment]
I'll leave this here for people to vote.

**Comment by webjib** [comment]
Although Webkit does not support standard JS API, there's a webkit API for managing Fullscreen (WebRTC custom card implements it):
https://developer.apple.com/documentation/webkitjs/htmlvideoelement/1633500-webkitenterfullscreen

I can clearly understand that this is only one manufacturer that does not implement the JS API, but we're talking about Apple, with a total market share of more than 20% for mobile phones :-)

**Comment by dermotduffy** [comment]
Thanks for the pointer @webjib !

**Comment by webjib** [comment]
> Thanks for the pointer [@webjib](https://github.com/webjib) !

I would love to help, but my knowledge in JS/TS is too weak :-(

**Comment by dermotduffy** [comment]
It looks like `webkitEnterFullscreen` only works on `HTMLVideoElement`, so that is not at all an equal substitute. It could make a single video element fullscreen, but I would assume iOS video player in the card already lets you make a single video fullscreen? That's different from being able to make the _card_ fullscreen -- which might have multiple videos, has the menu/status bar, or would have no effect at all in timeline/gallery since they have no video, or indeed some live providers which don't use video at all (e.g. `image` or `jsmpeg` live providers).

Is that still worth doing, even with those caveats?

**Comment by webjib** [comment]
I think this could help users of the card especially newbies. When I installed the card, I was surprised that the full screen button was there, but had no effect. Then I read your warning in the documentation but I was surprised because WebRTC card was implementing a fullscreen button.

I understand you would prefer a fullscreen that works for all features of the card, but maybe a fullscreen function only for the video player would help to watch a stream in a better way (it helps to watch the stream in landscape). Then you could add this limitation in the documentation: the fullscreen button is working only to make the active video player fullscreen, and not the whole card.

Would it make sense?

**Comment by dermotduffy** [comment]
> Would it make sense?

Yes, I think so -- I think it's clearer "better than nothing", so lets give it a shot!

**Comment by webjib** [comment]
You will make a lot of iOS users very happy!
Your card is so powerful, I've totally replaced my old WebRTC card ;-)

**Comment by Hukuma1** [comment]
Is it possible to get pinch-to-zoom to work alongside this fullscreen mode? Currently it seems to only work to zoom while not in fullscreen.

**Comment by dermotduffy** [comment]
> Is it possible to get pinch-to-zoom to work alongside this fullscreen mode? Currently it seems to only work to zoom while not in fullscreen.

I suspect not. The manufacturer in question has elected to implement very limited support for fullscreen from Javascript on iPhone iOS. It seems that any scaling transformations (e.g. to zoom in) have no effect. I don't have a device to actually test with, but from playing around in Browserstack the video element itself remains fixed size regardless of the scaling or other CSS applied, which is probably why it doesn't work.

**Comment by NickM-27** [comment]
it's worth mentioning that on iOS 18+ you can enable complete fullscreen API support in the safari features menu in settings

**Comment by dermotduffy** [comment]
> it's worth mentioning that on iOS 18+ you can enable complete fullscreen API support in the safari features menu in settings

Nice! And that works on iPhone? (The classic problem has been lack of Fullscreen API support in iOS on iPhone specifically, whilst it worked on iPad, as I understand it).

**Comment by NickM-27** [comment]
Correct

**Comment by webjib** [comment]
Are you sure it's not accessible only to Developpers? I can see this setting, but my iPhone is set up as a Developer device.

**Comment by NickM-27** [comment]
Yes, I am not an iOS developer and it is there under feature flags
